### PR TITLE
Restore predicate pushdown of metadata field in CheckpointEntryIterator

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

#18423 caused a regression about #17408
This is a short-term solution until https://github.com/trinodb/trino/issues/19156

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.

### CODE REVIEW

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java [hunk 1]
```diff
@@ -182,9 +183,20 @@ public CheckpointEntryIterator(
...
+        TupleDomain<HiveColumnHandle> tupleDomain;
+        if (columns.size() == 1) {
+            tupleDomain = buildTupleDomainColumnHandle(getOnlyElement(fields), getOnlyElement(columns));
+        }
+        else if (columns.size() == 2 && fields.contains(METADATA) && fields.contains(PROTOCOL)) {
+            // TODO https://github.com/trinodb/trino/issues/19156 Add support for predicate pushdown for both metadata and protocol in CheckpointEntryIterator
+            HiveColumnHandle metadata = columns.stream()
+                    .filter(column -> column.getBaseColumnName().equals("metadata"))
+                    .collect(onlyElement());
+            tupleDomain = buildTupleDomainColumnHandle(METADATA, metadata);
```
Comment (raunaqmorarka) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:195]: Is it the case that whenever "protocol" is non-null, then "metadata" is also guaranteed to be non-null ?
I'm wondering if this is safe given that this filter can exclude rows where "metadata" is null but "protocol" is non-null.

Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:195]: "metadata" is a required field and it's not nullable as far as I read Delta protocol.

Comment (raunaqmorarka) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:195]: sounds okay to me, would be great if someone more familiar with delta spec than me would also review this

Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:195]: @ebyhr the above quote doesn't actually specify that the _protocol_ and _metadata_ actions are in the same row group.
AFAIU, https://github.com/trinodb/trino/pull/17408 is about retrieving only the row groups which correspond to non-null filters.

For finding _protocol_ or _metadata_ there are different Parquet filters used:

https://github.com/trinodb/trino/blob/e1dc146f2409447318c47a0ebf1167fa4c315942/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java#L244-L251

In the absence of composed Parquet predicate pushdown, we're probably left with reading two times the checkpoint file (although this means reading actually two row groups - which are likely the same most of the time).

**Pull request closed without merging**